### PR TITLE
[soc] Silabs EFR32 enable DCDC before setting up clk

### PR DESCRIPTION
**Pull request by ekarlso**

### CODE REVIEW

File: soc/arm/silabs_exx32/Kconfig [hunk 1]
```diff
@@ -37,13 +42,14 @@ config HAS_CMU
 	  Set if the clock management unit (CMU) is present in the SoC.
 
 if HAS_CMU
-
+
```
Comment (galak) [inline soc/arm/silabs_exx32/Kconfig]: remove added whitespace.

File: soc/arm/silabs_exx32/Kconfig [hunk 2]
```diff
@@ -26,6 +26,11 @@ config SOC_PART_NUMBER
 	  that you should not set directly. The part number selection choice defines
 	  the default value for this string.
 
+config CONFIG_HAS_GECKO_DCDC
```
Comment (galak) [inline soc/arm/silabs_exx32/Kconfig]: should be 'config HAS_GECKO_DCDC'.  The CONFIG_... gets added by build system and is what's used in the source code.

File: soc/arm/silabs_exx32/Kconfig [hunk 3]
```diff
@@ -37,13 +42,14 @@ config HAS_CMU
...
 if HAS_CMU
-
+		
 choice
 	prompt "High Frequency Clock Selection"
 	default CMU_HFCLK_HFXO
 
 config CMU_HFCLK_HFXO
 	bool "External high frequency crystal oscillator"
+	select CONFIG_HAS_GECKO_DCDC
```
Comment (galak) [inline soc/arm/silabs_exx32/Kconfig]: should be HAS_GECKO_DCDC

**Review by galak — CHANGES_REQUESTED** [review]
Some cleanup of Kconfig symbols, also fixup the commit message to be "arm: silabs: ..."

**Review by galak — APPROVED** [review]

File: soc/arm/silabs_exx32/Kconfig [hunk 4]
```diff
@@ -26,6 +26,11 @@ config SOC_PART_NUMBER
 	  that you should not set directly. The part number selection choice defines
 	  the default value for this string.
 
+config HAS_GECKO_DCDC
+	bool
+	help
+		Enable the DCDC for the SoC.
```
Comment (MaureenHelm) [inline soc/arm/silabs_exx32/Kconfig]: Looks like the indentation is wrong here. Please indent one tab and two spaces.

I also think this config should be moved to `ext/hal/silabs/gecko/Kconfig`.

File: soc/arm/silabs_exx32/Kconfig [hunk 5]
```diff
@@ -36,14 +41,14 @@ config HAS_CMU
 	help
 	  Set if the clock management unit (CMU) is present in the SoC.
 
-if HAS_CMU
-
+if HAS_CMU
```
Comment (MaureenHelm) [inline soc/arm/silabs_exx32/Kconfig]: Please revert this unrelated change

File: soc/arm/silabs_exx32/common/soc.c [hunk 6]
```diff
@@ -70,6 +71,13 @@ static ALWAYS_INLINE void clkInit(void)
 #endif
 }
 
+static ALWAYS_INLINE void dcDcInit(void) {
+#ifdef CONFIG_HAS_GECKO_DCDC
+	EMU_DCDCInit_TypeDef init = EMU_DCDCINIT_DEFAULT;
+	EMU_DCDCInit(&init);
```
Comment (MaureenHelm) [inline soc/arm/silabs_exx32/common/soc.c]: Don't you also need a change in `ext/hal/silabs/gecko/CMakeLists.txt` to build the DCDC driver?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: ext/hal/silabs/gecko/Kconfig [hunk 7]
```diff
@@ -9,3 +9,8 @@ config HAS_SILABS_GECKO
 	bool
 	select HAS_CMSIS
 	depends on SOC_FAMILY_EXX32
+
+config HAS_GECKO_DCDC
```
Comment (mnkp) [inline ext/hal/silabs/gecko/Kconfig]: Not all projects will want to enable internal DC/DC converter. It has a limited power output and some applications will require an external one. This should be a visible option which is configurable and can be enabled by the board configuration file.

Let's rename it to 'GECKO_DCDC' or maybe 'SOC_GECKO_DCDC'.

File: ext/hal/silabs/gecko/Kconfig [hunk 8]
```diff
@@ -9,3 +9,8 @@ config HAS_SILABS_GECKO
...
+	bool
+	help
+      Enable the DCDC for the SoC.
```
Comment (mnkp) [inline ext/hal/silabs/gecko/Kconfig]: Missing end of line.

**Review by mnkp — CHANGES_REQUESTED** [review]
Thanks for the PR, I have some remarks though.

File: ext/hal/silabs/gecko/Kconfig (hunk 7 above)
Comment (ekarlso) [reply ext/hal/silabs/gecko/Kconfig]: I have added it to the board as a default config instead for it there

**Review by mnkp — CHANGES_REQUESTED** [review]
I had a closer look at all the different power configurations available on the gecko MCUs and I think we need to change our approach a bit as we won't be able to extend the current one to support all the cases.

First of all not all MCUs have a DCDC converter, e.g. efm32hg and efm32wg do not. In this case we could hide the SOC_DCDC_GECKO Kconfig option. Probably not a must but certainly nice to have. To do this one can use the 'depends on' but a better approach is the HAVE_ pattern as described in [Kconfig language](https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt) chapter "Adding common features and make the usage configurable".

The second issue is more important. I had a look at the reference manuals of the following two series EFR32xG1 and EFR32xG12. They support the following power configurations:
- initial (unconfigured)
- DC-DC off
- DC-DC on
- bypass mode

In case of EFR32xG1 the initial and bypass mode are identical. 

Having `SOC_DCDC_GECKO` option is not enough. It enables the DC/DC code but we need also the second option to choose in which mode the DC/DC converter should be configured. This can be achieved with a 'choice' statement as in
```
choice SOC_DCDC_GECKO_MODE
	prompt "DC/DC mode"
	depends on SOC_DCDC_GECKO
	help
	  Select power configuration mode of the on chip DC/DC converter.

	config SOC_DCDC_GECKO_MODE_UNCONFIGURED
		bool "Initial / Uconfigured"

	config SOC_DCDC_GECKO_MODE_ON
		bool "DC/DC On"

	config SOC_DCDC_GECKO_MODE_OFF
		bool "DC/DC Off"

	config SOC_DCDC_GECKO_MODE_BYPASS
		bool "Bypass"
endchoice
```
The default value will be `SOC_DCDC_GECKO_MODE_UNCONFIGURED` which is equivalent to not touching DC/DC at all.

In your case the board configuration file should select SOC_DCDC_GECKO and SOC_DCDC_GECKO_MODE_ON.

If you don't have time you don't need to implement all the features, the functionality can be added later when there is someone who needs it, but the Kconfig choice statement should be there.

File: soc/arm/silabs_exx32/Kconfig [hunk 9]
```diff
@@ -26,6 +26,12 @@ config SOC_PART_NUMBER
 	  that you should not set directly. The part number selection choice defines
 	  the default value for this string.
 
+config SOC_DCDC_GECKO
+	bool "Enable SOC DCDC regulator"
+	default n
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply soc/arm/silabs_exx32/Kconfig]: @mnkp but isnt UNCONFIGURED same as not doing any init at all basically? If mean if we never init it to on then ?

**Comment by ekarlso** [comment]
I pushed an update, @mnkp but it needs more work... Please elaborate on how you mean OFF ON and BYPASS to work together.

File: soc/arm/silabs_exx32/common/soc.c [hunk 10]
```diff
@@ -70,6 +71,25 @@ static ALWAYS_INLINE void clkInit(void)
 #endif
 }
 
+static ALWAYS_INLINE void dcDcInit(void)
```
Comment (mnkp) [inline soc/arm/silabs_exx32/common/soc.c]: I would code this function differently. Here is my proposal:
```
#ifdef CONFIG_SOC_GECKO_EMU_DCDC
static ALWAYS_INLINE void dcdc_init(void)
{
#if defined(CONFIG_SOC_GECKO_EMU_DCDC_MODE_UNCONFIGURED)
	/* Nothing to do, leave DC/DC converter in unconfigured, safe state. */
#elif defined(CONFIG_SOC_GECKO_EMU_DCDC_MODE_ON) || defined(CONFIG_SOC_GECKO_EMU_DCDC_MODE_BYPASS)
	EMU_DCDCInit_TypeDef init_cfg = EMU_DCDCINIT_DEFAULT;
#if defined(CONFIG_SOC_GECKO_EMU_DCDC_MODE_BYPASS)
	init_cfg.dcdcMode = emuDcdcMode_Bypass;
#endif
	EMU_DCDCInit(&init_cfg);
#elif defined(CONFIG_SOC_GECKO_EMU_DCDC_MODE_OFF)
	EMU_DCDCPowerOff();
#else
#error "Unsupported power configuration mode of the on chip DC/DC converter."
#endif
}
#endif
```
It should be called like this
```
#ifdef CONFIG_SOC_GECKO_EMU_DCDC
	/*
	 * Initialize DCDC before attempting to set clock, HFXO for example
	 * requires this
	 */
	dcdc_init();
#endif
```
Dependent code, e.g. Kconfig options should be updated accordingly.

Rationale:
- function name should follow Zephyr [Coding Style](https://docs.zephyrproject.org/latest/contribute/contribute_guidelines.html#coding-style). I understand that we already broke this rule and have names like clkInit, oldLevel. I reckon we should clean it up, albeit in a separate commit.
- surrounding call to `dcdc_init()` in the `#ifdef` is ugly, however makes it obvious for person reading `silabs_exx32_init()` function code that this part of the code is optional. Someone who's working with SiLabs MCU without DC/DC converter should not see a call to dcdc_init() function, even it the function is empty.
- let's not use `SOC_DCDC_GECKO` convention for SoC config options even if this is a style used by all the drivers e.g. `UART_GECKO_0`. In case of SoC options it makes them very confusing. Better convention to follow is `SOC_GECKO_<FEATURE>`. The addition of `_EMU` is optional but should make things easier when we add power management code which will also depend on the EMU module. You may remove the `_EMU` if you think it makes the name too long.

File: soc/arm/silabs_exx32/Kconfig [hunk 11]
```diff
@@ -26,6 +26,12 @@ config SOC_PART_NUMBER
...
+	help
+      Enable the DCDC for the SoC.
```
[the remark this answers is not in the document]
Comment (mnkp) [reply soc/arm/silabs_exx32/Kconfig]: Still not fixed.

File: soc/arm/silabs_exx32/Kconfig [hunk 12]
```diff
@@ -26,6 +26,30 @@ config SOC_PART_NUMBER
...
+		bool "DC/DC On"
+
+	config SOC_DCDC_GECKO_MODE_OFF
+		bool "DC/DC Off"
+
+	config SOC_DCDC_GECKO_MODE_BYPASS
+		bool "Bypass"
+endchoice
+
+config SOC_DCDC_GECKO
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: This option should select `HAS_EMU`. However `HAS_EMU` should be better renamed to `SOC_GECKO_EMU`.

File: soc/arm/silabs_exx32/Kconfig [hunk 13]
```diff
@@ -26,6 +26,30 @@ config SOC_PART_NUMBER
 	  that you should not set directly. The part number selection choice defines
 	  the default value for this string.
 
+choice SOC_DCDC_GECKO_MODE
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: This option should be placed below `SOC_DCDC_GECKO`.

File: soc/arm/silabs_exx32/Kconfig [hunk 14]
```diff
@@ -26,6 +26,30 @@ config SOC_PART_NUMBER
...
+
+	config SOC_DCDC_GECKO_MODE_OFF
+		bool "DC/DC Off"
+
+	config SOC_DCDC_GECKO_MODE_BYPASS
+		bool "Bypass"
+endchoice
+
+config SOC_DCDC_GECKO
+	bool "Enable SOC DCDC regulator"
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: Lets' change 'DCDC' to 'DC/DC'.

File: soc/arm/silabs_exx32/Kconfig [hunk 15]
```diff
@@ -26,6 +26,30 @@ config SOC_PART_NUMBER
...
+	help
+      Enable the DCDC for the SoC.
+
 config HAS_EMU
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: Let's change the name of this option to `SOC_GECKO_EMU`. It should be enabled when used, as it's the case with DCDC converter, not merely when EMU module is present in the SoC. The option will also be enabled by the power management code which we should add soon.

**Review by mnkp — CHANGES_REQUESTED** [review]
@ekarlso thanks for the update. We move in the right direction. I have one general remark though. I understand it's a work in progress however please test your code better before pushing to PR. The last version does not even compile, has spelling mistakes, some Kconfig options in .c code miss CONFIG_ prefix. 

Since we are touching common code we should also check that the code compiles on other SiLabs boards that support the feature and doesn't break those that don't.

**Comment by mnkp** [comment]
@galak, @MaureenHelm we're implementing the configuration of the on chip DC/DC regulator via Kconfig options. Since this is strictly boot related, board specific, not exposed to the user via any API I was wondering if it was not better to use DTS to do the job? We should take into account that DC/DC has a bunch of options, like output voltage, current limit, power mode which we don't expose currently but most likely we will need to support them in the future.

**Comment by ekarlso** [comment]
@mnkp any idea on how to do it with DTS?

**Comment by mnkp** [comment]
I'm not 100% sure we should use DTS. Until @galak or @MaureenHelm claim otherwise let's continue with Kconfig.

File: soc/arm/silabs_exx32/Kconfig (hunk 15 above)
Comment (galak) [reply soc/arm/silabs_exx32/Kconfig]: Do it as a separate commit.

**Comment by galak** [comment]
> I'm not 100% sure we should use DTS. Until @galak or @MaureenHelm claim otherwise let's continue with Kconfig.

I think we can do the DTS stuff later, not sure if treating this as a regulator makes sense:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/regulator/regulator.txt

**Comment by mnkp** [comment]
The DTS regulator defined by the kernel looks good, suitable for us.

**Comment by ekarlso** [comment]
@mnkp should we try to do Regulator now then or wait with it ?

**Review by MaureenHelm — DISMISSED** [review]
There are outstanding changes requested by @mnkp that still need to be addressed

File: soc/arm/silabs_exx32/Kconfig [hunk 16]
```diff
@@ -27,7 +27,32 @@ config SOC_PART_NUMBER
 	  that you should not set directly. The part number selection choice defines
 	  the default value for this string.
 
-config HAS_EMU
+choice SOC_GECKO_DCDC_MODE
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: We need to move this choice option below
```
config SOC_GECKO_EMU_DCDC
```
one. Otherwise the configuration interface available via menuconfig would be quite confusing. Also it should be renamed to 'SOC_GECKO_EMU_DCDC_MODE' since we decided to keep the 'EMU_'.

File: soc/arm/silabs_exx32/Kconfig [hunk 17]
```diff
@@ -27,7 +27,32 @@ config SOC_PART_NUMBER
...
+
+	config SOC_GECKO_EMU_DCDC_MODE_OFF
+		bool "DC/DC Off"
+
+	config SOC_GECKO_EMU_DCDC_MODE_BYPASS
+		bool "Bypass"
+endchoice
+
+config SOC_GECKO_EMU_DCDC
+	bool "Enable SOC DC/DC regulator"
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: It's customary written as: SoC

File: soc/arm/silabs_exx32/Kconfig [hunk 18]
```diff
@@ -27,7 +27,32 @@ config SOC_PART_NUMBER
...
+	select SOC_GECKO_EMU
+	help
+	  Enable the DC/DC for the SoC.
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: Better 'Enable the on chip DC/DC regulator' since 'DC/DC for the SoC' may also mean an external DC/DC. Or, since this text is not really different from the name of the option, we could get rid of the help text all together.

File: soc/arm/silabs_exx32/Kconfig [hunk 19]
```diff
@@ -27,7 +27,32 @@ config SOC_PART_NUMBER
...
+
+config SOC_GECKO_EMU
 	bool
 	help
 	  Set if the energy management unit (EMU) is present in the SoC.
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: better: 'Set if the energy management unit (EMU) is used.'

File: soc/arm/silabs_exx32/efr32mg12p/Kconfig.soc [hunk 20]
```diff
@@ -13,7 +13,7 @@ config SOC_EFR32MG12P
 	bool "SOC_EFR32MG12P"
 	select HAS_SILABS_GECKO
 	select HAS_CMU
-	select HAS_EMU
+	select SOC_GECKO_EMU
```
Comment (mnkp) [inline soc/arm/silabs_exx32/efr32mg12p/Kconfig.soc]: This should not be necessary any more. We should be able to remove this line (I did not test it).

**Review by mnkp — CHANGES_REQUESTED** [review]
Thanks @ekarlso. Looks good, I have a few minor comments. If we have a choice of implementing DTS regulator now or later I would choose later.

File: soc/arm/silabs_exx32/Kconfig (hunk 16 above)
Comment (mnkp) [reply soc/arm/silabs_exx32/Kconfig]: Let's change the name of this option to 'SOC_GECKO_EMU_DCDC_MODE', the 'EMU_' is missing.

File: soc/arm/silabs_exx32/common/soc.c [hunk 21]
```diff
@@ -92,6 +112,14 @@ static int silabs_exx32_init(struct device *arg)
 
 	_ClearFaults();
 
+#ifdef CONFIG_SOC_GECKO_EMU_DCDC
+	/*
+	 * Initialize DCDC before attempting to set clock, HFXO for example
```
Comment (mnkp) [inline soc/arm/silabs_exx32/common/soc.c]: As @chrta has noticed 'DCDC is not always required for HFXO. It depends on the board/schematic.' I agree with the statement. Let's maybe drop ', HFXO for example requires this' or otherwise we would need to go into too much detail to make this comment viable.

File: soc/arm/silabs_exx32/Kconfig [hunk 22]
```diff
@@ -29,10 +29,35 @@ config SOC_PART_NUMBER
...
+	  Enable the on chip DC/DC regulator
+
+choice SOC_GECKO_EMU_DCDC_MODE
+	prompt "DC/DC mode"
+	depends on SOC_GECKO_EMU_DCDC
+	help
+	  Select power configuration mode of the on chip DC/DC converter.
+
+	config SOC_GECKO_EMU_DCDC_MODE_UNCONFIGURED
+		bool "Initial / Uconfigured"
```
Comment (chrta) [inline soc/arm/silabs_exx32/Kconfig]: typo Uconfigured

**Review by chrta — CHANGES_REQUESTED** [review]

**Comment by mnkp** [comment]
@ekarlso would you mind pushing the last fix. It would be great to have this PR merged.

**Comment by mnkp** [comment]
@ekarlso thanks for the fix, however according to Zephyr's [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/contribute_guidelines.html#contribution-workflow) we need to use `git push --force ` when updating PRs and not follow up commits.

**Review by chrta — APPROVED** [review]

**Review by mnkp — APPROVED** [review]

**Comment by ekarlso** [comment]
@MaureenHelm can you take a look at this?

**Comment by ekarlso** [comment]
Can we fire off a recheck for this anyone ?

**Comment by mnkp** [comment]
An old comment from the devel mailing list:
> One can just post a comment to the PR saying ‘recheck’ to get CI to rebuild.
> -- galak

I believe the maintainers are currently busy/absent so it will take probably a few days before they can have a look at this PR and possibly merge it.

**Comment by ekarlso** [comment]
recheck

**Comment by mnkp** [comment]
> recheck

Interesting, it looks like it didn't work. Shippable still refers to the old log. I guess we need to wait for one of the maintainers.

**Comment by mnkp** [comment]
recheck

**Review by galak — APPROVED** [review]
